File: app/src/main/java/com/savvasdalkitsis/gameframe/control/view/ControlFragment.java
package com.savvasdalkitsis.gameframe.control.view;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.util.Log;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.SeekBar;
import android.widget.Spinner;
import android.widget.TextView;

import com.savvasdalkitsis.butterknifeaspects.aspects.BindLayout;
import com.savvasdalkitsis.gameframe.R;
import com.savvasdalkitsis.gameframe.control.presenter.ControlPresenter;
import com.savvasdalkitsis.gameframe.infra.navigation.Navigator;
import com.savvasdalkitsis.gameframe.infra.view.FragmentSelectedListener;
import com.savvasdalkitsis.gameframe.infra.view.Snackbars;
import com.savvasdalkitsis.gameframe.injector.infra.navigation.NavigatorInjector;
import com.savvasdalkitsis.gameframe.injector.presenter.PresenterInjector;
import com.savvasdalkitsis.gameframe.ip.model.IpAddress;
import com.savvasdalkitsis.gameframe.model.Brightness;
import com.savvasdalkitsis.gameframe.model.ClockFace;
import com.savvasdalkitsis.gameframe.model.CycleInterval;
import com.savvasdalkitsis.gameframe.model.DisplayMode;
import com.savvasdalkitsis.gameframe.model.PlaybackMode;
import com.shazam.android.aspects.base.fragment.AspectSupportFragment;

import butterknife.Bind;
import butterknife.OnClick;
import butterknife.OnItemSelected;

@BindLayout(R.layout.fragment_control)
public class ControlFragment extends AspectSupportFragment implements ControlView, FragmentSelectedListener {

    private final Navigator navigator = NavigatorInjector.navigator();
    private final ControlPresenter presenter = PresenterInjector.controlPresenter();

    @Bind(R.id.view_brightness)
    SeekBar brightness;
    @Bind(R.id.view_playback_mode)
    Spinner playbackMode;
    @Bind(R.id.view_cycle_interval)
    Spinner cycleInterval;
    @Bind(R.id.view_display_mode)
    Spinner displayMode;
    @Bind(R.id.view_clock_face)
    Spinner clockFace;
    @Bind(R.id.view_ip)
    TextView ip;
    @Bind(R.id.view_control_content)
    View content;
    @Bind(R.id.view_control_error)
    View error;
    FloatingActionButton fab;

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        fab = (FloatingActionButton) getActivity().findViewById(R.id.view_fab);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        brightness.setOnSeekBarChangeListener(new BrightnessChangedListener());
        playbackMode.setAdapter(adapter(R.array.playback_mode));
        cycleInterval.setAdapter(adapter(R.array.cycle_interval));
        displayMode.setAdapter(adapter(R.array.display_mode));
        clockFace.setAdapter(adapter(R.array.clock_face));
        presenter.bindView(this);
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.loadIpAddress();
    }

    @Override
    public void onFragmentSelected() {
        presenter.loadIpAddress();
        fab.setImageResource(R.drawable.ic_power_settings_new_white_48px);
        fab.setOnClickListener(v -> presenter.togglePower());
    }

    @Override
    public void onFragmentUnselected() {
    }

    @OnClick(R.id.view_menu)
    public void menu() {
        presenter.menu();
    }

    @OnClick(R.id.view_next)
    public void next() {
        presenter.next();
    }

    @OnClick(R.id.view_control_setup)
    public void setup() {
        navigator.navigateToIpSetup();
    }

    @OnClick(R.id.view_brightness_low)
    public void brightnessLow() {
        brightness.incrementProgressBy(-1);
    }

    @OnClick(R.id.view_brightness_high)
    public void brightnessHigh() {
        brightness.incrementProgressBy(1);
    }

    @OnItemSelected(R.id.view_playback_mode)
    public void playbackMode(int position) {
        presenter.changePlaybackMode(PlaybackMode.from(position));
    }

    @OnItemSelected(R.id.view_cycle_interval)
    public void cycleInterval(int position) {
        presenter.changeCycleInterval(CycleInterval.from(position));
    }

    @OnItemSelected(R.id.view_display_mode)
    public void displayMode(int position) {
        presenter.changeDisplayMode(DisplayMode.from(position));
    }

    @OnItemSelected(R.id.view_clock_face)
    public void clockFace(int position) {
        presenter.changeClockFace(ClockFace.from(position));
    }

    @Override
    public void operationSuccess() {
        Snackbars.success(getActivity().findViewById(R.id.view_coordinator), R.string.success).show();
    }

    @Override
    public void operationFailure(Throwable e) {
        Log.e(ControlFragment.class.getName(), "Operation failure", e);
        Snackbars.error(getActivity().findViewById(R.id.view_coordinator), R.string.operation_failed).show();
    }

    @Override
    public void ipAddressLoaded(IpAddress ipAddress) {
        ip.setText(String.format("Game Frame IP: %s", ipAddress.toString()));
        error.setVisibility(View.GONE);
        content.setVisibility(View.VISIBLE);
    }

    @Override
    public void ipCouldNotBeFound(Throwable throwable) {
        Log.e(ControlFragment.class.getName(), "Could not find ip", throwable);
        error.setVisibility(View.VISIBLE);
        content.setVisibility(View.GONE);
    }

    @NonNull
    private ArrayAdapter<CharSequence> adapter(int data) {
        ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(getContext(), data, android.R.layout.simple_spinner_item);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        return adapter;
    }

    private class BrightnessChangedListener implements SeekBar.OnSeekBarChangeListener {
        @Override
        public void onProgressChanged(SeekBar seekBar, int level, boolean b) {
            presenter.changeBrightness(Brightness.from(level));
        }

        @Override
        public void onStartTrackingTouch(SeekBar seekBar) {}

        @Override
        public void onStopTrackingTouch(SeekBar seekBar) {}
    }
}

File: app/src/main/java/com/savvasdalkitsis/gameframe/feature/control/view/ControlFragment.kt
/**
 * Copyright 2017 Savvas Dalkitsis
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * 'Game Frame' is a registered trademark of LEDSEQ
 */
package com.savvasdalkitsis.gameframe.feature.control.view

import android.os.Bundle
import android.support.annotation.ArrayRes
import android.support.design.widget.FloatingActionButton
import android.util.Log
import android.view.View
import android.widget.ArrayAdapter
import android.widget.SeekBar
import butterknife.OnClick
import butterknife.OnItemSelected
import com.savvasdalkitsis.gameframe.R
import com.savvasdalkitsis.gameframe.feature.control.model.*
import com.savvasdalkitsis.gameframe.feature.ip.model.IpAddress
import com.savvasdalkitsis.gameframe.infra.android.BaseFragment
import com.savvasdalkitsis.gameframe.infra.android.FragmentSelectedListener
import com.savvasdalkitsis.gameframe.infra.android.Snackbars
import com.savvasdalkitsis.gameframe.infra.kotlin.gone
import com.savvasdalkitsis.gameframe.infra.kotlin.visible
import com.savvasdalkitsis.gameframe.injector.presenter.PresenterInjector
import kotlinx.android.synthetic.main.fragment_control.*

class ControlFragment : BaseFragment(), ControlView, FragmentSelectedListener {

    private val presenter = PresenterInjector.controlPresenter()
    private lateinit var fab: FloatingActionButton

    override val layoutId: Int
        get() = R.layout.fragment_control

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        fab = activity.findViewById(R.id.view_fab_control)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        view_brightness.setOnSeekBarChangeListener(BrightnessChangedListener())
        view_playback_mode.adapter = adapter(R.array.playback_mode)
        view_cycle_interval.adapter = adapter(R.array.cycle_interval)
        view_display_mode.adapter = adapter(R.array.display_mode)
        view_clock_face.adapter = adapter(R.array.clock_face)
        presenter.bindView(this)
    }

    override fun onResume() {
        super.onResume()
        if (isVisible) {
            presenter.loadIpAddress()
        }
    }

    override fun onFragmentSelected() {
        presenter.loadIpAddress()
    }

    override fun onFragmentUnselected() {
        fab.gone()
    }

    @OnClick(R.id.view_menu)
    fun menu() {
        presenter.menu()
    }

    @OnClick(R.id.view_next)
    fun next() {
        presenter.next()
    }

    @OnClick(R.id.view_control_setup)
    fun setup() {
        presenter.setup()
    }

    @OnClick(R.id.view_brightness_low)
    fun brightnessLow() {
        view_brightness.incrementProgressBy(-1)
    }

    @OnClick(R.id.view_brightness_high)
    fun brightnessHigh() {
        view_brightness.incrementProgressBy(1)
    }

    @OnItemSelected(R.id.view_playback_mode)
    fun playbackMode(position: Int) {
        presenter.changePlaybackMode(PlaybackMode.from(position))
    }

    @OnItemSelected(R.id.view_cycle_interval)
    fun cycleInterval(position: Int) {
        presenter.changeCycleInterval(CycleInterval.from(position))
    }

    @OnItemSelected(R.id.view_display_mode)
    fun displayMode(position: Int) {
        presenter.changeDisplayMode(DisplayMode.from(position))
    }

    @OnItemSelected(R.id.view_clock_face)
    fun clockFace(position: Int) {
        presenter.changeClockFace(ClockFace.from(position))
    }

    override fun operationSuccess() = Snackbars.success(activity.findViewById(R.id.view_coordinator), R.string.success)

    override fun operationFailure(e: Throwable) {
        Log.e(ControlFragment::class.java.name, "Operation failure", e)
        Snackbars.error(activity.findViewById(R.id.view_coordinator), R.string.operation_failed)
    }

    override fun ipAddressLoaded(ipAddress: IpAddress) {
        view_ip.text = getString(R.string.game_frame_ip, ipAddress.toString())
        view_control_error.gone()
        view_control_content.visible()
        fab.visible()
        fab.setOnClickListener { presenter.togglePower() }
    }

    override fun ipCouldNotBeFound(throwable: Throwable) {
        Log.e(ControlFragment::class.java.name, "Could not find ip", throwable)
        view_control_error.visible()
        view_control_content.gone()
        fab.gone()
    }

    private fun adapter(@ArrayRes data: Int) =
            ArrayAdapter.createFromResource(context, data, android.R.layout.simple_spinner_item).apply {
                setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            }

    private inner class BrightnessChangedListener : SeekBar.OnSeekBarChangeListener {
        override fun onProgressChanged(seekBar: SeekBar, level: Int, b: Boolean) {
            presenter.changeBrightness(Brightness.from(level))
        }

        override fun onStartTrackingTouch(seekBar: SeekBar) {}

        override fun onStopTrackingTouch(seekBar: SeekBar) {}
    }
}
